### ISSUE TITLE
Resolve race condition in usage get_sendbuffer_ptr and send

### ISSUE THREAD

**Issue by sagarjha** [Derecho-Project/derecho-unified#44]
This has been a long standing problem, but we never ran into problem cases as we had just one thread sending messages in a subgroup in all test cases. We expect users to first get a buffer pointer from Derecho and then generate the message in it directly. And then call send to indicate to Derecho that the message has been generated and is ready to be sent. This is the pattern for sending uncooked (or raw) messages. For cooked messages, our library internally calls these functions when the user calls ordered_send/query etc. Thus the pattern of use that we support is `get_sendbuffer_ptr` and `send` one after the other.
While we do not expect the user to act as an adversary by calling `send` before `get_sendbuffer_ptr`, there is still a genuine race condition possible here. If there are multiple threads calling these functions, both threads can get a buffer one after the other. The buffers must be sent in order, so the send naturally sends the next buffer in the sender ring. If the thread that got the second buffer (in the order) calls `send` first, we will send the first buffer given to the other thread. That other thread might not have finished generating its message.

**Comment by sagarjha** [comment]
This was an easy fix. We just need to make sure that we never loan two buffers at the same time. If we gave a buffer to a thread, we must wait until send is called before we give out another buffer. We have enough data structures internally to track this. When we have loaned a buffer to a thread for a given subgroup, `next_sends[subgroup_num` will be non-null for an RDMC message and `pending_sst_sends[subgroup_num]` will be true for an SST message. So in `get_sendbuffer_ptr`, adding an additional check:

            if(pending_sst_sends[subgroup_num] || next_sends[subgroup_num]) {
                return nullptr;
           }

suffices. This makes sure that there are no SST or RDMC messages for which only the `send` call is pending. Reading and modifying these two data structures is well guarded inside locks everywhere, so there's no extra synchronization required.
